File: src/arcs/maze/mazeRecursion.cpp
#include <vector>
#include <cstdio>
#include <numeric>
#include <algorithm>
using namespace std;

int solution(vector<vector<int>> maze){
    if(maze.empty())
        return 0;

    if(maze.size() == 1)
        return accumulate(maze[0].begin(), maze[0].end(), 0);
    
    if(maze[0].size() == 1){
        auto res = 0;
        for(vector<int> pane : maze)
            res += pane[0];
        
        return res;
    }

    // 向右访问之后的子向量
    vector<vector<int>> rightSub;
    for(vector<int> row : maze){
        vector<int> everyRow {row.begin() + 1, row.end()};
        rightSub.push_back(everyRow);
    }

    // 向下访问之后的子向量
    vector<vector<int>> downSub {maze.begin() + 1, maze.end()};

    return maze[0][0] + min(solution(downSub), solution(rightSub));
}

int main(){

    vector<vector<int>> maze = {
        {1, 3, 1},
        {1, 5, 1},
        {4, 2, 1}
    };

    printf("%d\n", solution(maze));
    return 0;
}

File: src/arcs/maze/maze_recursion.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

def solution(maze):
    # 如果是空迷宫，直接返回0
    if not maze:
        return 0
    
    # 递归停止条件1：只有1行的迷宫
    if len(maze) == 1:
        return sum(maze[0])
    
    # 递归停止条件2：只有1列的迷宫
    if len(maze[0]) == 1:
        res = 0
        for pane in maze:
            res += pane[0]
        
        return res
    
    return maze[0][0] + min(solution(maze[1::]), solution([row[1::] for row in maze]))

if __name__ == '__main__':
    maze = [
        [1, 3, 1],
        [1, 5, 1],
        [4, 2, 1]
    ]

    print(solution(maze))